### ISSUE TITLE
Unable to redirect to capture.php after prepare.php is rendered

### ISSUE THREAD

**Issue by iadityakumar** [Payum/Payum#600]
I am trying to implement PayPal Express Checkout using  Payum and I am stuck at an error "The page isn’t redirecting properly Firefox has detected that the server is redirecting the request for this address in a way that will never complete." when I try to open prepare.php .
![screenshot from 2016-11-07 13-23-18](https://cloud.githubusercontent.com/assets/4803451/20050212/eb3b85ba-a4ed-11e6-8742-4cf566b1b374.png)

Below is my prepare and capture script.
```
<?php
// prepare.php

include_once 'config.php';

$gatewayName = 'paypal_express_checkout';

$storage = $payum->getStorage($paymentClass);

$payment = $storage->create();
$payment->setNumber(uniqid());
$payment->setCurrencyCode('EUR');


$payment->setDetails(array(
  // put here any fields in a gateway format.
  // for example if you use Paypal ExpressCheckout you can define a description of the first item:
  // 'L_PAYMENTREQUEST_0_DESC0' => 'A desc',
));

$storage->update($payment);

$captureToken = $payum->getTokenFactory()->createCaptureToken('paypal_express_checkout', $payment, 'done.php');

header("Location: ".$captureToken->getTargetUrl());
```

```
<?php
//capture.php

include_once 'config.php';


use Payum\Core\Reply\HttpPostRedirect;
use Payum\Core\Reply\HttpRedirect;
use Payum\Core\Request\Capture;
use Payum\Core\Security\TokenInterface;

$token = $payum->getHttpRequestVerifier()->verify($_REQUEST);
$gateway = $payum->getGateway($token->getGatewayName());
if ($reply = $gateway->execute(new Capture($token), true)) {
    if ($reply instanceof HttpRedirect) {
        header("Location: ".$reply->getUrl());
        exit();
    }
    if ($reply instanceof HttpPostRedirect) {
        echo($reply->getContent());
        exit();
    }
    throw new \LogicException('Unsupported reply', null, $reply);
}
$payum->getHttpRequestVerifier()->invalidate($token);
header("Location: ".$token->getAfterUrl());
```

**Comment by makasim** [comment]
As far as I can see the redirect url is relative for the current page. Try to make it absolute. Try to change the line (add a slash):

```
header("Location: ".$captureToken->getTargetUrl());
```

to 

```
header("Location: ".'/'.$captureToken->getTargetUrl());
```

**Comment by iadityakumar** [comment]
It is again giving some other error 
![screenshot from 2016-11-07 14-25-18](https://cloud.githubusercontent.com/assets/4803451/20051678/4e823fd0-a4f6-11e6-9a21-a2685bbfd490.png)

**Comment by makasim** [comment]
the url is not correct it must be `payumimp/capture.php?payum_token=xxxx`

**Comment by makasim** [comment]
Try to change it manually

**Comment by iadityakumar** [comment]
Now this error message is coming up : 

Fatal error: Uncaught Payum\Core\Exception\InvalidArgumentException: The current url /payumimp/capture.php?payum_token=9LxdXDp-q4t6_09TC6I8Uz4efCtcrGPbLJ4YRW7QHdI not match target url http://localhost/payumimp/prepare.php/capture.php/capture.php?payum_token=9LxdXDp-q4t6_09TC6I8Uz4efCtcrGPbLJ4YRW7QHdI set in the token. in /opt/lampp/htdocs/payumimp/vendor/payum/core/Payum/Core/Bridge/PlainPhp/Security/HttpRequestVerifier.php:55 Stack trace: #0 /opt/lampp/htdocs/payumimp/capture.php(14): Payum\Core\Bridge\PlainPhp\Security\HttpRequestVerifier->verify(Array) #1 {main} thrown in /opt/lampp/htdocs/payumimp/vendor/payum/core/Payum/Core/Bridge/PlainPhp/Security/HttpRequestVerifier.php on line 55.

**Comment by iadityakumar** [comment]
This url "payumimp/capture.php?payum_token=xxxx" must be generated automatically right .

**Comment by makasim** [comment]
Looks like there is a bug in https://github.com/Payum/Payum/blob/master/src/Payum/Core/Bridge/PlainPhp/Security/TokenFactory.php#L31 method. Could you try to debug it?

**Comment by freddurst1805** [comment]
UP ! I got the exact same error

**Comment by freddurst1805** [comment]
It's pretty tough to get it fixed alone. So please can you help ?

**Comment by iadityakumar** [comment]
I tried to debug the code but was unable to do it.

**Comment by richardwo** [comment]
I'm having the same problem. Following the steps in the 'Get it started' guide results in this error. Chrome displays the message 'This page isn’t working ERR_TOO_MANY_REDIRECTS'

**Comment by ponasromas** [comment]
The bug is still not resolved. Any solutions?

**Comment by makasim** [comment]
> The bug is still not resolved. Any solutions?

The solution is to go ahead and fix it. I am sure the problem is in this method https://github.com/Payum/Payum/blob/master/src/Payum/Core/Bridge/PlainPhp/Security/TokenFactory.php#L31 which is not taking into account path part of the url.

**Comment by tpk109** [comment]
I could not find a way to get the needed dirname from HttpUri, so I fixed it like this:

`protected function generateUrl($path, array $parameters = [])
    {
        //$newPath = (new Path($this->baseUrl->getPath()))->withTrailingSlash().$path;
        $newPath = (new Path($this->baseUrl->getPath()))->withTrailingSlash();
        $pathinfo = pathinfo($newPath);

        // If we are in root of server, dirname is empty and Windows returns for dirname a "\"
        $dirname = str_replace("\\", "/", $pathinfo["dirname"]);

        // If we are not in root but domain/payum/prepare.php, dirname returns "/payum", so add slash
        $newPath = $dirname.'/'.$path;

        // Remove double slash, if we are in root
        $newPath = str_replace("//", "/", $newPath);

        $uri = $this->baseUrl->withPath($newPath);
        $uri = $this->addQueryToUri($uri, $parameters);


        return (string) $uri;
    }`

I am just trying to work myself into Payum. As I understand this is a very basic function, as the generated uris are checked in the other files. As this is not fixed since ages I wonder what that means for the whole project?

**Comment by pierredup** [comment]
This has been fixed in #889
